internal/app/machined/pkg/controllers/runtime/maintenance_service.go

```diff
@@ -0,0 +1,275 @@
+			if err != nil && !state.IsNotFoundError(err) {
+				return fmt.Errorf("failed to get machine config: %w", err)
+			}
+
+			if state.IsNotFoundError(err) {
+				if err = r.Create(ctx, configResource); err != nil {
+					return fmt.Errorf("failed to create machine config: %w", err)
+				}
+			} else {
+				configResource.Metadata().SetVersion(oldConfigResource.Metadata().Version())
+
+				if err = configResource.Metadata().SetOwner(oldConfigResource.Metadata().Owner()); err != nil {
+					return fmt.Errorf("error setting owner: %w", err)
+				}
+
+				if err = r.Update(ctx, configResource); err != nil {
+					return fmt.Errorf("failed to update machine config: %w", err)
+				}
+			}
+
+			continue
+		case <-r.EventCh():
+		}
+
+		request, err := safe.ReaderGetByID[*runtime.MaintenanceServiceRequest](ctx, r, runtime.MaintenanceServiceRequestID)
+		if err != nil && !state.IsNotFoundError(err) {
+			return fmt.Errorf("failed to get maintenance service request: %w", err)
+		}
+
+		if request == nil {
+			// no request, nothing to do
+			shutdownServer(ctx)
+
+			continue
+		}
+
+		if request.Metadata().Phase() == resource.PhaseTearingDown {
+			// stop the server & remove the finalizer
+			shutdownServer(ctx)
+
+			if err = r.RemoveFinalizer(ctx, request.Metadata(), ctrl.Name()); err != nil {
+				return fmt.Errorf("failed to remove finalizer: %w", err)
+			}
+
+			continue
+		}
+
+		cfg, err := safe.ReaderGetByID[*runtime.MaintenanceServiceConfig](ctx, r, runtime.MaintenanceServiceConfigID)
+		if err != nil && !state.IsNotFoundError(err) {
+			return fmt.Errorf("failed to get maintenance service config: %w", err)
+		}
+
+		cert, err := safe.ReaderGetByID[*secrets.MaintenanceServiceCerts](ctx, r, secrets.MaintenanceServiceCertsID)
+		if err != nil && !state.IsNotFoundError(err) {
+			return fmt.Errorf("failed to get maintenance service certs: %w", err)
+		}
+
+		if cert != nil {
+			if err = tlsProvider.Update(cert); err != nil {
+				return fmt.Errorf("failed to update tls provider: %w", err)
+			}
+		}
+
+		// immediately add a finalizer
+		if err = r.AddFinalizer(ctx, request.Metadata(), ctrl.Name()); err != nil {
+			return fmt.Errorf("failed to add finalizer: %w", err)
+		}
+
+		if cfg == nil {
+			// no config, nothing to do
+			shutdownServer(ctx)
+
+			continue
+		}
+
+		if listener == nil {
+			listener, err = net.Listen("tcp", cfg.TypedSpec().ListenAddress)
+			if err != nil {
+				return fmt.Errorf("failed to listen: %w", err)
+			}
+		}
+
+		if server == nil {
+			tlsConfig, err := tlsProvider.TLSConfig()
+			if err != nil {
+				return fmt.Errorf("failed to get tls config: %w", err)
+			}
+
+			server = factory.NewServer(
+				srv,
+				factory.WithDefaultLog(),
+				factory.ServerOptions(
+					grpc.Creds(
+						credentials.NewTLS(tlsConfig),
+					),
+				),
+
+				factory.WithUnaryInterceptor(injector.UnaryInterceptor()),
+				factory.WithStreamInterceptor(injector.StreamInterceptor()),
+			)
+
+			go func() {
```

Comment:
can we add that to errgoup and do `eg.Wait()` in `shutdownServer` to be on the save side?

Reply:
I can look into that, need to make sure we can re-use the waitgroup, as the controller might start the server multiple times

Reply:
fixed, added `sync.WaitGroup`